### PR TITLE
Listen to LSP connection file changes

### LINKED ISSUE #243: Listen to file changes and update the index
Listen to file changes and update the index:
- delete
- rename
- create?

see connection.onDidChangeWatchedFiles

### PR DESCRIPTION
**Pull request by msujew**
Closes #243

Also fixes an issue related to the glob pattern which was used in the yeoman generator. The generated pattern was incorrect and changes were never picked up.

You can test this by opening the `domain-model` example workspace and **only** open one of the files in vscode. Deleting/Changing files in the workspace outside of vscode should invalidate references. Changing them back or restoring them should create valid files again.

### CODE REVIEW

**Comment by dhuebner** [comment]
I checked the domain model example and can verify that it works. 

I have an additional question about FS watcher. Will the watcher created with `workspace.createFileSystemWatcher` be disposed automatically when client.stop() is called or is it not necessary at all?

File: packages/langium/src/documents/document-builder.ts [hunk 1]
```diff
@@ -76,40 +77,54 @@ export class DefaultDocumentBuilder implements DocumentBuilder {
...
         return diagnostics;
     }
 
-    async documentChanged(uri: URI, cancelToken = CancellationToken.None): Promise<void> {
-        this.langiumDocuments.invalidateDocument(uri);
-        const newDocument = this.langiumDocuments.getOrCreateDocument(uri);
+    async update(changed: URI[], deleted: URI[], cancelToken = CancellationToken.None): Promise<void> {
+        // Gather deleted documents first, as they won't be available later
+        // They are needed to correctly collect all affected documents
+        const deletedDocuments = deleted.map(e => this.langiumDocuments.getOrCreateDocument(e));
```
Comment (spoenemann) [inline packages/langium/src/documents/document-builder.ts:92]: AFAICT we only need the URIs of the deleted documents, not their LangiumDocument instances.

File: packages/langium/src/index/index-manager.ts [hunk 2]
```diff
@@ -23,26 +23,33 @@ export interface IndexManager {
...
-    update(document: LangiumDocument | LangiumDocument[], cancelToken?: CancellationToken): Promise<void>;
+    update(documents: LangiumDocument[], cancelToken?: CancellationToken): Promise<void>;
 
     /**
      * Returns all documents that could be affected by the specified document.
      *
      * @param document The document which may affect other documents.
      */
-    getAffectedDocuments(document: LangiumDocument): Stream<LangiumDocument>;
+    getAffectedDocuments(documents: LangiumDocument[]): Stream<LangiumDocument>;
```
Comment (spoenemann) [inline packages/langium/src/index/index-manager.ts:52]: Use the URIs as input so we don't have to resolve deleted Documents.

File: packages/langium/src/lsp/language-server.ts [hunk 3]
```diff
@@ -82,23 +81,33 @@ export function startLanguageServer(services: LangiumServices): void {
...
-        } catch (err) {
-            if (err !== OperationCancelled) {
-                throw err;
-            }
-        }
+        changePromise = documentBuilder
+            .update(changed, deleted ?? [], changeTokenSource.token)
+            .catch(err => {
+                if (err !== OperationCancelled) {
+                    console.log('Error: ', err);
```
Comment (spoenemann) [inline packages/langium/src/lsp/language-server.ts:99]: Log errors to `console.error`

File: packages/langium/src/lsp/language-server.ts [hunk 4]
```diff
@@ -82,23 +81,33 @@ export function startLanguageServer(services: LangiumServices): void {
...
-        try {
-            changePromise = documentBuilder.documentChanged(URI.parse(change.document.uri), changeTokenSource.token);
-        } catch (err) {
-            if (err !== OperationCancelled) {
-                throw err;
-            }
-        }
+        changePromise = documentBuilder
+            .update(changed, deleted ?? [], changeTokenSource.token)
+            .catch(err => {
```
Comment (spoenemann) [inline packages/langium/src/lsp/language-server.ts:97]: Why did you change to a callback instead of a `catch` block?

**Review by spoenemann — COMMENTED** [review]
@dhuebner you are right. FileSystemWatcher extends Disposable, so we should add all created watchers to `context.subscriptions` so they're cleaned up when the extension is stopped.

File: packages/langium/src/lsp/language-server.ts (hunk 4 above)
Comment (msujew) [reply packages/langium/src/lsp/language-server.ts:97]: Because the `catch` block doesn't actually catch the exception there (since we don't await the promise). It only catches it if we are at the previous `await` instruction, but there's no guarantee that we actually hit that code.

File: packages/generator-langium/langium-template/src/extension.ts [hunk 5]
```diff
@@ -34,12 +33,15 @@ function startLanguageClient(context: vscode.ExtensionContext): LanguageClient {
         debug: { module: serverModule, transport: TransportKind.ipc, options: debugOptions }
     };
 
+    const fileSystemWatcher = vscode.workspace.createFileSystemWatcher('**/*.<%= file-glob-extension %>');
+	context.subscriptions.push(fileSystemWatcher);
```
Comment (spoenemann) [inline packages/generator-langium/langium-template/src/extension.ts:37]: Nitpick: inconsistent whitespace

**Review by spoenemann — APPROVED** [review]
:+1:

File: packages/generator-langium/langium-template/src/extension.ts (hunk 5 above)
Comment (msujew) [reply packages/generator-langium/langium-template/src/extension.ts]: Thanks, somehow a tab character found it's way into there 🤔
